src/StreamFactoryInterface.php

```diff
@@ -9,15 +9,11 @@
     /**
      * Create a new stream from a resource or a string.
      *
-     * If the argument is a resource, it MUST be readable and SHOULD be seekable. It MAY be writable.
-     * If the argument is a string, a temporary resource will be created that is writable and seekable.
-     * 
-     * If the arugment is a string it will be interpided as the content of the stream. File names or 
-     * file paths are not supported. 
+     * String content will be used as the body of the stream. Files paths are not supported.
```

Comment:
@shadowhand 

> The problem with providing a file path as the target is two-fold

I agree, and actually the constructor is overloaded an inconsistent in Diactoros - it's actually two constructors rolled into one:

```php
__construct(string $path, string $mode = "r");
__construct(resource $handle);
```

Even if the resulting streams are the same type, the factory method signatures are obviously completely different - which reinforces my point about separate factories.

Whether the resulting stream type is the same for those two methods is quite irrelevant to me - the ways they get created are different, which to me implies different factories, even if the underlying implementations both happen to use the same overloaded constructor; that's an implementation detail, which may not be true of every implementation.

> It is one function call for users to open a file and pass it as a resource and at least two function calls for users to create a resource from string content

I understand that's not as *practical* as what I'm suggesting, but still feel like the distinction is important: wrapping an existing resource in a stream, and *creating* a memory-based resource from contents, are two distinct operations to me. Wrapping an existing resource, and opening and wrapping an existing resource, I find, are much more closely related operations.

I still think the problem is best addressed by three distinct interfaces and methods - none of which should be named `createStream`, as it's too non-specific.

For these three cases, I'd suggest:

```php
FileStreamFactory::createFileStream(string $path, string $mode = "r");
ResourceStreamFactory::createResourceStream(resource $handle);
TempStreamFactory::createTempStream(string $content);
MemoryStreamFactory::createMemoryStream(string $content);
```

Note that I'm proposing two distinct factories for `php://temp` and `php://memory` based streams - one is suitable for smaller chunks, the other more suitable for writing large streams such as data-exports.

As these are arguably the most important types, in terms of interoperability, I'd suggest these be part of the core standard, not a supplemental package or standard. Needing to create a stream from these common types of PHP and OS resources is extremely common - and in almost any ordinary case, if you need a stream, you're going to need *one* of those types of streams.

If somebody wants the utility of a "magic" factory that creates a stream from a variable/unknown type of resource, nothing stops them from doing that:

```php
class CommonStreamFactory
{
    private $fs, $rs;

    public function __construct(
        FileStreamFactory $fs,
        ResourceStreamFactory $rs
    ) {
        $this->fs = $fs;
        $this->rs = $rs;
    }

    /**
     * @param string|resource $path_or_handle
     * @param string $mode
     *
     * @return StreamInterface
     */
    public function createStream($path_or_handle, $mode = "r") {
        return is_resource($path_or_handle)
            ? $this->rs->createResourceStream($path_or_handle)
            : $this->fs->createFileStream($path_or_handle, $mode);
    }
}
```

My point is, this is a matter of convenience (and opinions) and not about interoperability - the interfaces minimally provide that *without* being opinionated, dictating implementation details, or making assumptions (or mandates) about the underlying implementations.

You can *have* all the convenience and utility you want - by simply installing packages. The PSR itself doesn't need to concern itself with ease of use. It should have a strong, unbiased focus on defining and isolating responsibilities - packages can then implement all the abstraction and utility you could wish for, in a manner where architectural differences do not become a liability, by allowing for difference of opinions, in the form of interoperable packages providing actual and *different* facilities.

An interface is not a facility - it's a contract for an eco-system of facilities.

@Nyholm @shadowhand @http-interop/http-factory-contributors I feel this may be the most important point I've contributed to this discussion so far, so please read carefully and think on it guys? :-)